File: src/Go/pipeline_services/core/registry.go
package core

import "fmt"

type PipelineRegistry[ContextType any] struct {
	actions      map[string]any
	unaryActions map[string]func(ContextType) ContextType
}

func NewPipelineRegistry[ContextType any]() *PipelineRegistry[ContextType] {
	return &PipelineRegistry[ContextType]{
		actions:      map[string]any{},
		unaryActions: map[string]func(ContextType) ContextType{},
	}
}

func (registry *PipelineRegistry[ContextType]) RegisterUnary(
	name string,
	action func(ContextType) ContextType,
) {
	if action == nil {
		panic("action must not be nil")
	}
	registry.unaryActions[name] = action
}

func (registry *PipelineRegistry[ContextType]) RegisterAction(name string, action any) {
	if action == nil {
		panic("action must not be nil")
	}
	// Validate eagerly while retaining the original callable shape for the
	// Pipeline's one normalization path.
	_ = normalizeAction[ContextType](action)
	registry.actions[name] = action
}

func (registry *PipelineRegistry[ContextType]) HasUnary(name string) bool {
	_, found := registry.unaryActions[name]
	return found
}

func (registry *PipelineRegistry[ContextType]) HasAction(name string) bool {
	_, found := registry.actions[name]
	return found
}

func (registry *PipelineRegistry[ContextType]) GetUnary(
	name string,
) (func(ContextType) ContextType, error) {
	action, found := registry.unaryActions[name]
	if !found {
		return nil, fmt.Errorf("unknown unary Action: %s", name)
	}
	return action, nil
}

func (registry *PipelineRegistry[ContextType]) GetAction(
	name string,
) (any, error) {
	action, found := registry.actions[name]
	if !found {
		return nil, fmt.Errorf("unknown Action: %s", name)
	}
	return action, nil
}

File: src/Rust/src/core/registry.rs
use std::collections::HashMap;
use std::sync::Arc;

use crate::core::pipeline::{ActionControl, StepAction, UnaryOperator};

pub struct PipelineRegistry<ContextType> {
  unary_actions: HashMap<String, UnaryOperator<ContextType>>,
  step_actions: HashMap<String, StepAction<ContextType>>,
}

impl<ContextType> PipelineRegistry<ContextType> {
  pub fn new() -> Self {
    Self {
      unary_actions: HashMap::new(),
      step_actions: HashMap::new(),
    }
  }

  pub fn register_unary<ActionFn>(&mut self, name: impl Into<String>, action: ActionFn)
  where
    ActionFn: Fn(ContextType) -> ContextType + Send + Sync + 'static,
  {
    self.unary_actions.insert(name.into(), Arc::new(action));
  }

  pub fn register_action<ActionFn>(&mut self, name: impl Into<String>, action: ActionFn)
  where
    ActionFn: Fn(ContextType, &mut ActionControl<ContextType>) -> ContextType + Send + Sync + 'static,
  {
    self.step_actions.insert(name.into(), Arc::new(action));
  }

  pub fn has_unary(&self, name: &str) -> bool {
    self.unary_actions.contains_key(name)
  }

  pub fn has_action(&self, name: &str) -> bool {
    self.step_actions.contains_key(name)
  }

  pub fn get_unary(&self, name: &str) -> Result<UnaryOperator<ContextType>, String> {
    match self.unary_actions.get(name) {
      Some(action) => Ok(action.clone()),
      None => Err(format!("Unknown unary action: {name}")),
    }
  }

  pub fn get_action(&self, name: &str) -> Result<StepAction<ContextType>, String> {
    match self.step_actions.get(name) {
      Some(action) => Ok(action.clone()),
      None => Err(format!("Unknown step action: {name}")),
    }
  }
}

impl<ContextType> Default for PipelineRegistry<ContextType> {
  fn default() -> Self {
    Self::new()
  }
}
